fv3core/grid/geometry.py

```diff
@@ -0,0 +1,905 @@
+
+        # ew = np.stack((ew1, ew2), axis=-1)
+
+        # fill ghost on ew:
+
+        if tile_partitioner.on_tile_left(rank):
+            if tile_partitioner.on_tile_bottom(rank):
+                _fill_ghost(ew1, 0.0, nhalo, "sw")
+                _fill_ghost(ew2, 0.0, nhalo, "sw")
+            if tile_partitioner.on_tile_top(rank):
+                _fill_ghost(ew1, 0.0, nhalo, "nw")
+                _fill_ghost(ew2, 0.0, nhalo, "nw")
+        if tile_partitioner.on_tile_right(rank):
+            if tile_partitioner.on_tile_bottom(rank):
+                _fill_ghost(ew1, 0.0, nhalo, "se")
+                _fill_ghost(ew2, 0.0, nhalo, "se")
+            if tile_partitioner.on_tile_top(rank):
+                _fill_ghost(ew1, 0.0, nhalo, "ne")
+                _fill_ghost(ew2, 0.0, nhalo, "ne")
+
+    else:
+        ew1[:, :, 1] = 1.0
+        ew2[:, :, 2] = 1.0
+        # ew = np.stack((ew1, ew2), axis=-1)
+
+    return ew1[1:-1, :, :], ew2[1:-1, :, :]
+
+
+def calc_unit_vector_south(
+    xyz_dgrid, xyz_agrid, grid_type, nhalo, tile_partitioner, rank, np
+):
+    """
+    Calculates the cartesian unit vector pointing south from every grid cell.
+    The first set of values is the horizontal component, the second is the vertical
+    component as defined by the cell edges -- in a non-spherical grid these will be
+    x and y unit vectors.
+    """
+    es1 = np.zeros((xyz_agrid.shape[0], xyz_dgrid.shape[1], 3))
+    es2 = np.zeros((xyz_agrid.shape[0], xyz_dgrid.shape[1], 3))
+    if grid_type < 3:
+
+        pp = xyz_midpoint(xyz_dgrid[:-1, 1:-1, :3], xyz_dgrid[1:, 1:-1, :3])
+        p2 = np.cross(xyz_agrid[:, :-1, :3], xyz_agrid[:, 1:, :3])
+        if tile_partitioner.on_tile_bottom(rank):
+            p2[:, nhalo - 1] = np.cross(pp[:, nhalo - 1], xyz_agrid[:, nhalo, :3])
+        if tile_partitioner.on_tile_top(rank):
+            p2[:, -nhalo] = np.cross(xyz_agrid[:, -nhalo - 1, :3], pp[:, -nhalo])
+
+        es2[:, 1:-1, :] = normalize_xyz(np.cross(p2, pp))
+
+        p1 = np.cross(xyz_dgrid[:-1, 1:-1, :], xyz_dgrid[1:, 1:-1, :])
+        es1[:, 1:-1, :] = normalize_xyz(np.cross(p1, pp))
+
+        # es = np.stack((es1, es2), axis=-1)
+
+        # fill ghost on es:
+        if tile_partitioner.on_tile_left(rank):
+            if tile_partitioner.on_tile_bottom(rank):
+                _fill_ghost(es1, 0.0, nhalo, "sw")
+                _fill_ghost(es2, 0.0, nhalo, "sw")
+            if tile_partitioner.on_tile_top(rank):
+                _fill_ghost(es1, 0.0, nhalo, "nw")
+                _fill_ghost(es2, 0.0, nhalo, "nw")
+        if tile_partitioner.on_tile_right(rank):
+            if tile_partitioner.on_tile_bottom(rank):
+                _fill_ghost(es1, 0.0, nhalo, "se")
+                _fill_ghost(es2, 0.0, nhalo, "se")
+            if tile_partitioner.on_tile_top(rank):
+                _fill_ghost(es1, 0.0, nhalo, "ne")
+                _fill_ghost(es2, 0.0, nhalo, "ne")
```

Comment:
There are a lot of duplicated lines here with the "west" routine, can you consolidate them?

I thought there is a `tile_partitioner.on_tile_bottom_left` or something similar, if not we should add one later when we consolidate names to use west/east/north/south.

Reply:
I didn't see anything like `on_tile_bottom_left` in the partitioner, but I agree it would be a good idea to add something along those lines